test/e2e/volumes.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// Start a client pod using given VolumeSource (exported by startVolumeServer())
+// and check that the pod sees the data from the server pod.
+func testVolumeClient(client *client.Client, config VolumeTestConfig, volume api.VolumeSource, expectedContent string) {
+	By(fmt.Sprint("starting ", config.prefix, " client"))
+	podClient := client.Pods(config.namespace)
+
+	clientPod := &api.Pod{
+		TypeMeta: api.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1beta3",
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name: config.prefix + "-client",
+			Labels: map[string]string{
+				"role": config.prefix + "-client",
+			},
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:       config.prefix + "-client",
+					Image:      "gcr.io/google_containers/nginx:1.7.9",
+					Privileged: false,
+					Ports: []api.ContainerPort{
+						{
+							Name:          "web",
+							ContainerPort: 80,
+							Protocol:      api.ProtocolTCP,
+						},
+					},
+					VolumeMounts: []api.VolumeMount{
+						{
+							Name:      config.prefix + "-volume",
+							MountPath: "/usr/share/nginx/html",
+						},
+					},
+				},
+			},
+			Volumes: []api.Volume{
+				{
+					Name:         config.prefix + "-volume",
+					VolumeSource: volume,
+				},
+			},
+		},
+	}
+	if _, err := podClient.Create(clientPod); err != nil {
+		Failf("Failed to create %s pod: %v", clientPod.Name, err)
+	}
+	expectNoError(waitForPodRunningInNamespace(client, clientPod.Name, config.namespace))
+
+	By("reading a web page from the client")
+	body, err := client.Get().
+		Namespace(config.namespace).
+		Prefix("proxy").
+		Resource("pods").
+		Name(clientPod.Name).
+		DoRaw()
+	expectNoError(err, "Cannot read web page: %v", err)
+	Logf("body: %v", string(body))
+
+	By("checking the page content")
+	Expect(body).To(ContainSubstring(expectedContent))
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// NFS
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-NFS", func() {
+	clean := true // If 'false', the test won't clear its namespace (and pods) upon completion. Useful for debugging.
+
+	// filled in BeforeEach
+	var c *client.Client
+	var namespace *api.Namespace
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+		By("Building a namespace api object")
+		namespace, err = createTestingNS("volume-nfs", c)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		if err := c.Namespaces().Delete(namespace.Name); err != nil {
+			Failf("Couldn't delete ns %s", err)
+		}
+	})
+
+	It("should be able to mount NFS", func() {
+		config := VolumeTestConfig{
+			namespace:   namespace.Name,
+			prefix:      "nfs",
+			serverImage: "gcr.io/google_containers/volume-nfs",
+			serverPorts: []int{2049},
+		}
+
+		defer func() {
+			if clean {
+				volumeTestCleanup(c, config)
+			}
+		}()
+
+		pod := startVolumeServer(c, config)
```

Comment:
To be clear, otherwise you will leave pods lying around, which will eventually fill up the cluster (e.g. for soak tests which run the e2e tests repeatedly against a long-running cluster).